### PR TITLE
Update middleware to document constructor vs. Invoke DI and scoping

### LINKED ISSUE #2031: Document lifetime of services in middleware
See https://github.com/aspnet/Mvc/issues/3860

### PR DESCRIPTION
**Pull request by ardalis**
Fixes #2031

### CODE REVIEW

File: aspnetcore/fundamentals/middleware.md [hunk 1]
```diff
@@ -147,7 +147,7 @@ RequestLoggerMiddleware.cs
...
 
 
-The middleware follows the [Explicit Dependencies Principle](http://deviq.com/explicit-dependencies-principle/) and exposes all of its dependencies in its constructor. Middleware can take advantage of the `UseMiddleware<T>` extension to inject services directly into their constructors, as shown in the example below. Dependency injected services are automatically filled, and the extension takes a `params` array of arguments to be used for non-injected parameters.
+The middleware follows the [Explicit Dependencies Principle](http://deviq.com/explicit-dependencies-principle/) and exposes its dependencies in its constructor. Middleware is constructed once per *application lifetime* - if you need to share services with middleware within a request, see *Per-request dependencies* below. Middleware can take advantage of the `UseMiddleware<T>` extension to inject services directly into their constructors, as shown in the example below. Dependency injected services are automatically filled, and the extension takes a `params` array of arguments to be used for non-injected parameters.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/middleware.md]: Consider:

Middleware follows the [Explicit Dependencies Principle](http://deviq.com/explicit-dependencies-principle/) and exposes its dependencies in its constructor. Middleware is constructed once per *application lifetime*. See *Per-request dependencies* below if you need to share services with middleware within a request. Use the `UseMiddleware<T>` extension to inject services directly into their constructors (shown in the example below).

File: aspnetcore/fundamentals/middleware.md [hunk 2]
```diff
@@ -166,6 +166,28 @@ Testing the middleware (by setting the `Hosting:Environment` environment variabl
 > [!NOTE]
 > The `UseStaticFiles` extension method (which creates the `StaticFileMiddleware`) also uses `UseMiddleware<T>`. In this case, the `StaticFileOptions` parameter is passed in, but other constructor parameters are supplied by `UseMiddleware<T>` and dependency injection.
 
+### Per-request dependencies
+
+Because middleware are constructed at app startup, not per-request, *transient* and *scoped* lifetime services used by middleware constructors will not be shared with other dependency-injected types during each request. If you need to share a *scoped* service between your middleware and other types, you should inject add these services to the `Invoke` method's signature. The `Invoke` method can accept additional parameters that will be populated by dependency injection. For example:
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/middleware.md]: >you should ~~inject~~ add these

Comment (ardalis) [inline aspnetcore/fundamentals/middleware.md]: Maybe remove transient to avoid confusion.

**Comment by ardalis** [comment]
@Rick-Anderson should be ready to go now

**Pull request merged**